### PR TITLE
[distill][phase1-2] Decouple DMD2 from Wan + YAML training args + checkpoint/resume

### PR DESCRIPTION
**Pull request by alexzms**
## 1) Motivation

Phase 0 #1120  introduced a new distillation scaffold (`Trainer ↔ Method ↔ Adapter + ModelBundle`), but it still had two big limitations:

- **Algorithm/model coupling still leaked through** (e.g. Wan-specific method naming and pipeline-backed behavior).
- **Entrypoints and configs were still “legacy-shaped”**, which makes it hard to scale to many models/methods/roles without re-creating a new `*_distillation_vN.py` per model family.

This PR lands **Phase 1 + Phase 2**, pushing the refactor to the point where we can run **few-step distillation via a YAML-only entrypoint** and keep the **method/algorithm reusable**, while the **adapter absorbs model/pipeline quirks**.

---

### What Phase 1 changes
- Move distillation logic toward a **FastGen-style hierarchy** where:
  - **`DistillTrainer`** is infra-only (loop/accum/step/logging).
  - **`DistillMethod`** owns the algorithm and update policy (multi-optimizer schedules, stepping cadence).
  - **`DistillAdapter`** owns model/pipeline-specific forward context and batch normalization.

### Key outcomes
- **Generic algorithm method**: introduce a reusable DMD2 implementation under a method taxonomy:
  - `fastvideo/distillation/methods/distribution_matching/dmd2.py`
- **Model-family adapter**: Wan specifics live in the adapter (forward context, pipeline normalization):
  - `fastvideo/distillation/adapters/wan.py`
- **Validation boundary**: validation is no longer a “pipeline side-effect”; it becomes an explicit component rather than being baked into legacy pipelines.

---

## 3) Phase 2: YAML-only entrypoint + builder runtime + fully decouple

Phase 2 makes the new path **standalone** (no legacy distillation pipeline dependency):

### New entrypoint (YAML-only)
- `fastvideo/training/distillation.py`
  - Only accepts **new YAML configs** (no legacy config fallback / merging).
  - CLI stays minimal: runtime controls like `--config`, `--resume-from-checkpoint`, `--override-output-dir`, `--dry-run`.

Example YAML:
``` yaml
distill:
  model: wan
  method: dmd2

models:
  student:
    family: wan
    path: Wan-AI/Wan2.1-T2V-1.3B-Diffusers
    trainable: true
  teacher:
    family: wan
    path: Wan-AI/Wan2.1-T2V-14B-Diffusers
    trainable: false
  critic:
    family: wan
    path: Wan-AI/Wan2.1-T2V-1.3B-Diffusers
    trainable: true

training:
  # Distributed
  num_gpus: 8
  sp_size: 1
  tp_size: 1

  # Data (parquet dataset folder)
  data_path: data/Wan-Syn_77x448x832_600k
  dataloader_num_workers: 4
  ...
```

### YAML config loading + typed spec
- `fastvideo/distillation/yaml_config.py`
- `fastvideo/distillation/specs.py`

### Builder that instantiates runtime from roles + method
- `fastvideo/distillation/builder.py`
  - Builds roles (student/teacher/critic/…) → adapter → model bundle → method → trainer.
  - Keeps “how to assemble a run” separate from both trainer and method.

### Validation without legacy pipelines
- `fastvideo/distillation/validators/base.py`
- `fastvideo/distillation/validators/wan.py`
  - Phase 2 runs validation via the new validator path (no legacy `_log_validation`).

### Checkpoint save/resume (new system)
- `fastvideo/distillation/checkpoint.py`
  - Save/resume trainable roles, optimizer/scheduler, dataloader state, and RNG states (including adapter-exposed generators).
  - Integrated into `fastvideo/distillation/trainer.py`.

### “outside/” config tree (non-invasive)
- `fastvideo/distillation/outside/fastvideo/configs/distillation/...`
  - Used to iterate on a better distillation config scheme without touching `fastvideo/configs/*` yet.
  - This PR’s entrypoint reads explicit paths; no “outside path auto-completion” magic.

---

### ✅ Done in this PR (Phase 1 + 2)
- Land a **generic DMD2 method** in a our new distillation framework.
- Land a **model-family adapter** (`WanAdapter`) that owns forward context (keeps methods clean).
- Add a **YAML-only distillation entrypoint**: `fastvideo/training/distillation.py`.
- Add a **role-based builder** that instantiates adapter/bundle/method/trainer from YAML.
- Add **independent validation** (no legacy distillation pipeline dependency).
- Add **checkpoint save/resume** to the new path (with retention policy).
- Provide a runnable few-step distillation config + example script:
  - YAML: `fastvideo/distillation/outside/fastvideo/configs/distillation/distill_wan2.1_t2v_1.3B_dmd2_8steps.yaml`
  - Runner: `examples/distillation/phase2/temp.sh` (Will be cleaned in phase 3)

### ❌ Not done yet (planned follow-ups)
- Additional distillation methods beyond DMD2 (e.g. Self-Forcing + ODE init path).
- Additional adapters/models beyond Wan.
- Long-term cleanup: remove legacy distillation pipelines (left intact for now; new code path is additive).
- More rigorous automated tests for end-to-end training/validation (GPU CI, SSIM regression, etc.).
- Since finetuning is a special distillation (only `student` and dataset would be provided), our new distillation framework is fully capable of aborbing the finetuning code into itself, resulting in a even cleaner training pipeline.
- Better config yaml design. For supporting future distillation methods or even finetuning.
- Decouple builder function in builer.py

---

## Tests / Evidence
- Manual GPU runs (few-step Wan DMD2) show the expected behavior: **early validation outputs are blurry and become progressively clearer and higher quality as training proceeds**.
- Phase 1,2 training behavior matches Phase 0 very closely in practice.
   - Phase 1: https://wandb.ai/alexzms-ucsd/phase1_wan_dmd2_8steps_wansyn/runs/dlyw3gh3?nw=nwuseralexzms
   - Phase 2: https://wandb.ai/alexzms-ucsd/phase2_wan_dmd2_8steps_wansyn/runs/77zfy2r2?nw=nwuseralexzms
---

### Feedback and suggestions are highly welcome!

### CODE REVIEW

File: fastvideo/distillation/validators/wan.py [hunk 1]
```diff
@@ -0,0 +1,250 @@
+# SPDX-License-Identifier: Apache-2.0
```
Comment (jzhang38) [inline fastvideo/distillation/validators/wan.py:1]: Most things in this file is model agnostic

File: fastvideo/distillation/adapters/wan.py [hunk 2]
```diff
@@ -0,0 +1,767 @@
+# SPDX-License-Identifier: Apache-2.0
```
Comment (jzhang38) [inline fastvideo/distillation/adapters/wan.py:1]: Most thing in this file is still very specific to DMD2, and agnostic to wan?

Comment (alexzms) [reply fastvideo/distillation/adapters/wan.py:1]: That's right. I will fix it

**Comment by alexzms** [comment]
Up to this version, only validator is not decoupled and decoupling will require to rewrite WanPipeline to accept mode of sde/ode. Planned to be done during phase3.

**Comment by alexzms** [comment]
https://wandb.ai/alexzms-ucsd/phase2.9_wan_dmd2_8steps_wansyn/runs/2rwoplxd
